handler.go

```diff
@@ -23,170 +26,159 @@ import (
+		if len(ca.Config.HubUser) == 0 {
+			return newEmptyFieldError(fieldHubUser)
+		} else if len(ca.Config.HubPassword) == 0 {
+			return newEmptyFieldError(fieldHubPassword)
+		}
+		err := fmt.Errorf("unable to authorize with provided Hub credentials (HTTP %d)", resp.StatusCode)
+		return err
+	} else if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusBadRequest {
+		err := fmt.Errorf("unable to authorize with provided Hub credentials (HTTP %d)", resp.StatusCode)
+		return err
 	}
+	return nil
+}
 
-	if resp.StatusCode == 401 && ca.Config.HubUser == "" {
-		return fmt.Errorf("Unable to authorise without credentials. Please set 'HUB USER' and 'HUB PASSWORD' from your Cloudradar account")
-	} else if resp.StatusCode == 401 && ca.Config.HubUser != "" {
-		return fmt.Errorf("Unable to authorise with provided credentials. Please correct 'HUB USER' and 'HUB PASSWORD' according to your Cloudradar account")
-	} else if resp.StatusCode < 200 || resp.StatusCode >= 400 {
-		return fmt.Errorf("Got bad response status: %d, %s. If you have a proxy or firewall it may be blocking the connection", resp.StatusCode, resp.Status)
-	}
+func newEmptyFieldError(name string) error {
+	err := fmt.Errorf("unexpected empty field %s", name)
+	return errors.Wrap(err, "the field must be filled with details of your Cloudradar account")
+}
 
-	return nil
+func newFieldError(name string, err error) error {
+	return errors.Wrapf(err, "%s field verification failed", name)
 }
 
-func (ca *Cagent) PostResultsToHub(result Result) error {
-	ca.initHubHTTPClient()
+func (ca *Cagent) PostResultToHub(ctx context.Context, result *Result) error {
+	ca.initHubClientOnce()
 
+	if len(ca.Config.HubURL) == 0 {
+		return newEmptyFieldError("hub_url")
+	} else if u, err := url.Parse(ca.Config.HubURL); err != nil {
+		return newFieldError("hub_url", err)
+	} else if u.Scheme != "http" && u.Scheme != "https" {
+		err := fmt.Errorf("wrong scheme '%s', URL must start with http:// or https://", u.Scheme)
+		return newFieldError("hub_url", err)
+	}
 	b, err := json.Marshal(result)
 	if err != nil {
+		err = errors.Wrap(err, "failed to serialize result")
 		return err
 	}
 
 	var req *http.Request
-
 	if ca.Config.HubGzip {
-		var buffer bytes.Buffer
-		zw := gzip.NewWriter(&buffer)
-		zw.Write(b)
-		zw.Close()
-		req, err = http.NewRequest("POST", ca.Config.HubURL, &buffer)
+		buf := new(bytes.Buffer)
+		gzipped := gzip.NewWriter(buf)
+		if _, err := gzipped.Write(b); err != nil {
+			err = errors.Wrap(err, "failed to write into gzipped buffer")
+			return err
+		}
+		if err := gzipped.Close(); err != nil {
+			err = errors.Wrap(err, "failed to finalize gzipped buffer")
+			return err
+		}
+		req, _ = http.NewRequest("POST", ca.Config.HubURL, buf)
 		req.Header.Set("Content-Encoding", "gzip")
 	} else {
-		req, err = http.NewRequest("POST", ca.Config.HubURL, bytes.NewBuffer(b))
+		req, _ = http.NewRequest("POST", ca.Config.HubURL, bytes.NewBuffer(b))
```

Comment:
Can we ignore this `err`?

Reply:
`NewRequest` may return error in these cases:
* Invalid method (it's hardcoded here)
* Invalid URL (but it's checked in the code above)

So given that both parameters are valid, there will be no expected error.

Reply:
> So given that both parameters are valid, there will be no expected error.

Well, I would still prefer an explicit error checking... who knows how the code might be changed in the future.

Reply:
@soupdiver if the code will change to some extent that `http.NewRequest` will fail, we will need to panic or fatal here, because it's critical to initialize a new request. This will trigger a E2E error in CI one day.

If someone will be changing this particular function call, he may add error checking as well, that will match his scenario.

Reply:
> he may add error checking as well, that will match his scenario.

People are lazy... if the error is omitted once, it will probably stay like that... because it was done like this before.

I already wasted hours of my life because I omitted errors "that should not happen" but for some reason still did.

If `http.NewRequest` returns an error for some reason, we end up here: https://github.com/cloudradar-monitoring/cagent/blob/4972b16a28779fb2a595458fe57ca233470f06ce/handler.go#L353 and post an error message. I think this is better than making an assumption that the case will never happen